llvm/lib/Target/NVPTX/NVPTXIncreaseAlignment.cpp

```diff
@@ -0,0 +1,131 @@
+//===-- NVPTXIncreaseAlignment.cpp - Increase alignment for local arrays --===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// A simple pass that looks at local memory arrays that are statically
+// sized and sets an appropriate alignment for them. This enables vectorization
+// of loads/stores to these arrays if not explicitly specified by the client.
+//
+// TODO: Ideally we should do a bin-packing of local arrays to maximize
+// alignments while minimizing holes.
+//
+//===----------------------------------------------------------------------===//
+
+#include "NVPTX.h"
+#include "llvm/IR/DataLayout.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/MathExtras.h"
+
+using namespace llvm;
+
+static cl::opt<bool>
+    MaxLocalArrayAlignment("nvptx-use-max-local-array-alignment",
+                           cl::init(false), cl::Hidden,
+                           cl::desc("Use maximum alignment for local memory"));
```

Comment:
We may as well allow the option to specify the exact alignment, instead of a boolean knob with a vague max value.

Reply:
OK. How about the upper limit on the alignment. E.g. the pass may want to align float3 by 16, but the user wants to avoid/minimize  the gaps and limit the alignment of 4 or 8?

Reply:
Sounds good. I've switched the option to one where the user may specify a minimum alignment and the pass will raise all allocas to that alignment except in cases where it doesn't make sense (specified alignment is bigger than something we could load, alloca is small enough that it can be loaded/stored via a single instruction at a lower alignment). Setting this to 16 will get the previous aggressive behavior and setting it to 1 will get the previous conservative behavior.